### PR TITLE
UX + Security current user password reset

### PR DESCRIPTION
**Pull request by coolaj86**
Follow up to https://github.com/go-gitea/gitea/pull/5034

I'll annotate these a little later, but it's combos for these states:

* not logged in
* logged in as expected user
* logged in as an unexpected user
* valid code
* invalid code (or expired code)

Default Condition
------------------

* Not logged In
* Valid Reset Code

Result

* Shows email of user for whom the password is to be reset
* Shows "Remember me"
* Signs in the user

screen shot 2018-10-07 at 11 15 55 am

Logged In
----------

* Already Logged In
* Valid Reset Code

Result

* Shows email of user for whom the password is to be reset
* Doesn't show "Remember me"
* Renews user session

screen shot 2018-10-07 at 11 15 31 am

Invalid Code
----------

* Login Status N/A
* Invalid Code

Result

* Error about invalid code
* Submits to same error page (could probably just disable submit)

screen shot 2018-10-07 at 11 14 22 am

Invalid User
----------

* User Logged in as FOO USER
* Valid Code for BAR USER

Result

* Error about mismatch user
* Submits to same error page (could probably just disable submit)

screen shot 2018-10-07 at 11 13 35 am

screen shot 2018-10-07 at 11 13 49 am

### CODE REVIEW

File: routers/user/auth.go [hunk 1]
```diff
@@ -1135,64 +1139,89 @@ func ForgotPasswdPost(ctx *context.Context) {
...
+	}
+
+	return u
+}
+
+// ResetPasswd render the reset password page
+func ResetPasswd(ctx *context.Context) {
 	ctx.Data["IsResetForm"] = true
+
+	_ = commonResetPassword(ctx)
```
Comment (lafriks) [inline routers/user/auth.go]: no need to assign if you don't need any response from function, just use `commonResetPassword(ctx)`

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/user/auth.go (hunk 1 above)
Comment (coolaj86) [reply routers/user/auth.go]: I understand that it's not needed, but I thought that this was a good way to improve readability - to show to the reader that I'm aware that there is a return, but that I explicitly don't need to use it.

Do you still want this changed?

Comment (lafriks) [reply routers/user/auth.go]: No need for extra code :) Please remove `_ = ` part

**Review by daviian — CHANGES_REQUESTED** [review]
I'm against auto-signin on password reset. Consequently I find having "Remember me" on Password reset awkward.

IMO we need a "password confirmation" field instead.
Showing Email is fine.

**Comment by coolaj86** [comment]
@daviian The "Remember me" felt a little weird to me too.

What I came to was that it's actually the form was incorrectly labeled. "Password Reset" is the process that happens on the Account page. This process is actually "Account Recovery", which is a different user flow.

The user isn't coming to this page with the intent to change their password (that is already on the Account page), but rather to be able to login to their account. In this case, with the current configuration, that implicitly includes setting a password. However, once more secure options are available (such as simply omitting passwords if a mailer is configured) the user will still want to recover their account (i.e. their primary oauth is down and they can't log in) but a password won't be part of the process.

I'll post new screenshots a little later.

File: routers/routes/routes.go [hunk 2]
```diff
@@ -185,6 +185,10 @@ func RegisterRoutes(m *macaron.Macaron) {
 	m.Get("/^:type(issues|pulls)$", reqSignIn, user.Issues)
 
 	// ***** START: User *****
+	m.Group("/user", func() {
```
[the remark this answers is not in the document]
Comment (coolaj86) [reply routers/routes/routes.go]: fixed

**Comment by coolaj86** [comment]
Two more things that should change for increased clarity:

* [ ] "New Password" instead of "Password"
* [ ] "Remember this device" instead of "Remember me"

**Comment by solderjs** [comment]
I'm back on the scene. I'll see what I can do to get this stuff cleaned up and PR'd again.

**Comment by solderjs** [comment]
Rebased and ready for re-review!

**Review by lafriks — APPROVED** [review]

File: routers/routes/routes.go [hunk 3]
```diff
@@ -273,8 +273,6 @@ func RegisterRoutes(m *macaron.Macaron) {
 		}, openIDSignInEnabled)
 		m.Get("/sign_up", user.SignUp)
 		m.Post("/sign_up", bindIgnErr(auth.RegisterForm{}), user.SignUpPost)
-		m.Get("/reset_password", user.ResetPasswd)
```
Comment (adelowo) [inline routers/routes/routes.go:276]: Do we need to atleast redirect to the new route before this is finally removed in 1.10?

**Review by adelowo — APPROVED** [review]

File: routers/routes/routes.go (hunk 3 above)
Comment (lafriks) [reply routers/routes/routes.go:276]: @adelowo I don't see why as emails having old route won't be usable anyway as code is valid only for few hours

**Comment by lafriks** [comment]
@daviian need your approval

File: routers/routes/routes.go (hunk 3 above)
Comment (adelowo) [reply routers/routes/routes.go:276]: Ha true... stupid me

**Review by daviian — APPROVED** [review]

**Pull request merged**